Firestore/core/src/firebase/firestore/core/firestore_client.mm

```diff
@@ -0,0 +1,485 @@
+using core::ViewSnapshot;
+using local::LruParams;
+using model::DatabaseId;
+using model::Document;
+using model::DocumentKeySet;
+using model::DocumentMap;
+using model::MaybeDocument;
+using model::Mutation;
+using model::OnlineState;
+using remote::Datastore;
+using remote::RemoteStore;
+using util::Path;
+using util::AsyncQueue;
+using util::DelayedConstructor;
+using util::DelayedOperation;
+using util::Executor;
+using util::Status;
+using util::StatusCallback;
+using util::StatusOr;
+using util::StatusOrCallback;
+using util::TimerId;
+
+std::shared_ptr<FirestoreClient> FirestoreClient::Create(
+    const DatabaseInfo& database_info,
+    const api::Settings& settings,
+    std::shared_ptr<auth::CredentialsProvider> credentials_provider,
+    std::shared_ptr<util::Executor> user_executor,
+    std::shared_ptr<util::AsyncQueue> worker_queue) {
+  // Have to use `new` because `make_shared` cannot access private constructor.
+  std::shared_ptr<FirestoreClient> client(
+      new FirestoreClient(database_info, std::move(credentials_provider),
+                          std::move(user_executor), std::move(worker_queue)));
+  client->Initialize(settings);
+  return client;
+}
+
+FirestoreClient::FirestoreClient(
+    const DatabaseInfo& database_info,
+    std::shared_ptr<auth::CredentialsProvider> credentials_provider,
+    std::shared_ptr<util::Executor> user_executor,
+    std::shared_ptr<util::AsyncQueue> worker_queue)
+    : database_info_(database_info),
+      credentials_provider_(std::move(credentials_provider)),
+      worker_queue_(std::move(worker_queue)),
+      user_executor_(std::move(user_executor)) {
+}
+
+void FirestoreClient::Initialize(const api::Settings& settings) {
+  if (client_initialized_) {
+    return;
+  }
+  auto user_promise = std::make_shared<std::promise<User>>();
+  bool credentials_initialized = false;
+
+  std::weak_ptr<FirestoreClient> weak_self(shared_from_this());
+  auto credential_change_listener = [credentials_initialized, user_promise,
+                                     weak_self](User user) mutable {
+    auto shared_self = weak_self.lock();
+    if (!shared_self) return;
+
+    if (!credentials_initialized) {
+      credentials_initialized = true;
+      user_promise->set_value(user);
+    } else {
+      shared_self->worker_queue()->Enqueue([shared_self, user] {
+        shared_self->worker_queue()->VerifyIsCurrentQueue();
+
+        LOG_DEBUG("Credential Changed. Current user: %s", user.uid());
+        [shared_self->sync_engine_ credentialDidChangeWithUser:user];
+      });
+    }
+  };
+
+  credentials_provider_->SetCredentialChangeListener(
+      credential_change_listener);
+
+  // Defer initialization until we get the current user from the
+  // credential_change_listener. This is guaranteed to be synchronously
+  // dispatched onto our worker queue, so we will be initialized before any
+  // subsequently queued work runs.
+  auto shared_self = shared_from_this();
```

Comment:
Consider moving this declaration above the declaration of `weak_self`, above and using it as the argument to the weak_self constructor.

Reply:
No longer relevant as it is moved to factory function.